### PR TITLE
feat: ✨ Add datahaven native transfer precompile

### PR DESCRIPTION
**Pull request by gonzamontiel**
## DataHaven Native Transfer Precompile

Implements EVM precompile at address `0x00000000000000000000000000000007F5` (2073) to expose `pallet-datahaven-native-transfer` functionality to the EVM layer.

### Features
- **Transfer to Ethereum**: Locks native tokens and sends them via Snowbridge to Ethereum addresses
- **View Functions**: Query paused state, total locked balance, and sovereign account address

### Implementation
- Precompile using `#[precompile_utils::precompile]` macro with proper gas accounting
- 15+ test cases covering success/failure scenarios
- Solidity interface with NatSpec documentation for contract integration

Enables seamless cross-chain transfers of DataHaven native tokens to Ethereum L1.

### CODE REVIEW

**Comment by stiiifff** [comment]
As discussed, not sure the Pause/Unpause operations should be exposed in the precompile. Otherwise, looks good.

File: operator/precompiles/datahaven-native-transfer/DataHavenNativeTransfer.sol [hunk 1]
```diff
@@ -0,0 +1,70 @@
+// SPDX-License-Identifier: GPL-3.0-only
+pragma solidity >=0.8.3;
+
+/// @dev The DataHavenNativeTransfer precompile address.
+address constant DATAHAVEN_NATIVE_TRANSFER_ADDRESS = 0x00000000000000000000000000000000000007f5;
```
Comment (ahmadkaouk) [inline operator/precompiles/datahaven-native-transfer/DataHavenNativeTransfer.sol]: `0x00000000000000000000000000000000000007F5` corresponds to 2037 which is not the address of the precompile used in the runtime.
https://github.com/datahaven-xyz/datahaven/blob/b76e13d911e346d0279af1d9a3af99e9169edf00/operator/runtime/mainnet/src/precompiles.rs#L145-L149

File: operator/precompiles/datahaven-native-transfer/DataHavenNativeTransfer.sol [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+/// @dev The DataHavenNativeTransfer precompile instance.
+DataHavenNativeTransfer constant DATAHAVEN_NATIVE_TRANSFER_CONTRACT =
+    DataHavenNativeTransfer(DATAHAVEN_NATIVE_TRANSFER_ADDRESS);
+
+/// @author The DataHaven Team
+/// @title DataHaven Native Transfer Interface
+/// @notice Interface for transferring DataHaven native tokens to/from Ethereum via Snowbridge
+/// @custom:address 0x00000000000000000000000000000000000007F5
```
Comment (ahmadkaouk) [inline operator/precompiles/datahaven-native-transfer/DataHavenNativeTransfer.sol]: Should fix the address

File: operator/precompiles/datahaven-native-transfer/README.md [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+This precompile exposes the `pallet-datahaven-native-transfer` functionality to the EVM layer, allowing smart contracts to transfer DataHaven native tokens to Ethereum via Snowbridge.
+
+## Overview
+
+The DataHaven Native Transfer precompile provides an EVM-compatible interface for:
+- Transferring native tokens from DataHaven to Ethereum
+- Managing the pallet's operational state (pause/unpause)
+- Querying transfer statistics and system state
+
+**Precompile Address:** `0x00000000000000000000000000000000000007F5` (2073 decimal)
```
Comment (ahmadkaouk) [inline operator/precompiles/datahaven-native-transfer/README.md]: Fix the address

File: operator/precompiles/datahaven-native-transfer/src/tests.rs [hunk 4]
```diff
@@ -0,0 +1,759 @@
...
+#[test]
+fn test_view_functions_gas_costs() {
+    ExtBuilder::default()
+        .with_balances(vec![(EthereumSovereign.into(), 1000)])
+        .build()
+        .execute_with(|| {
+            // isPaused should have minimal gas cost
+            precompiles()
+                .prepare_test(Alice, precompile_address(), PCall::is_paused {})
+                .expect_cost(0) // TODO: Calculate actual expected cost
```
Comment (ahmadkaouk) [inline operator/precompiles/datahaven-native-transfer/src/tests.rs]: What’s the deal with this TODO?

File: operator/precompiles/datahaven-native-transfer/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+        handle.record_cost(RuntimeHelper::<Runtime>::db_write_gas_cost())?;
+
+        // Convert caller address to substrate account
+        let caller = Runtime::AddressMapping::into_account_id(handle.context().caller);
+
+        // Build the call
+        let call = NativeTransferCall::<Runtime>::pause {}.into();
+
+        // Dispatch the call - this will fail if caller doesn't have permission
+        RuntimeHelper::<Runtime>::try_dispatch(handle, Some(caller).into(), call, 0)?;
```
Comment (ahmadkaouk) [inline operator/precompiles/datahaven-native-transfer/src/lib.rs]: Thinking again about this, this call will always revert. The precompile dispatches with a signed origin while the runtime requires `EnsureRoot` for `PauseOrigin`. No EVM caller can produce Root as far as I know, so the always revert with `BadOrigin`.

It's probably better to remove the pause and unpause calls.

File: operator/precompiles/datahaven-native-transfer/DataHavenNativeTransfer.sol [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+    /// @param from The account initiating the transfer
+    /// @param to The Ethereum address receiving the tokens
+    /// @param amount The amount of tokens transferred
+    event TokensTransferredToEthereum(address indexed from, address indexed to, uint256 amount);
+
+    /// @notice Emitted when the pallet is paused
+    event Paused();
+
+    /// @notice Emitted when the pallet is unpaused
+    event Unpaused();
```
Comment (ahmadkaouk) [inline operator/precompiles/datahaven-native-transfer/DataHavenNativeTransfer.sol]: We are declaring events here but the precompile Rust implementation never constructs a log or calls record_log. So no EVM receipt logs are produced.

File: operator/precompiles/datahaven-native-transfer/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+    /// - `amount`: Amount of tokens to transfer (in smallest unit)
+    /// - `fee`: Fee to incentivize relayers (in smallest unit)
+    #[precompile::public("transferToEthereum(address,uint256,uint256)")]
+    fn transfer_to_ethereum(
+        handle: &mut impl PrecompileHandle,
+        recipient: Address,
+        amount: U256,
+        fee: U256,
+    ) -> EvmResult {
+        // Ensure we're not in a static context
```
Comment (ahmadkaouk) [inline operator/precompiles/datahaven-native-transfer/src/lib.rs]: What does this comment mean?

File: operator/precompiles/datahaven-native-transfer/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+        handle.record_cost(RuntimeHelper::<Runtime>::db_read_gas_cost())?;
```
Comment (ahmadkaouk) [inline operator/precompiles/datahaven-native-transfer/src/lib.rs]: Why are we recording a db_read_gas_cost here?

File: operator/precompiles/datahaven-native-transfer/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,228 @@
...
+    }
+
+    /// Unpause the pallet
+    ///
+    /// Allows token transfers again after being paused.
+    /// Only callable by accounts with the PauseOrigin permission.
+    #[precompile::public("unpause()")]
+    fn unpause(handle: &mut impl PrecompileHandle) -> EvmResult {
+        // Record db write cost
+        handle.record_cost(RuntimeHelper::<Runtime>::db_write_gas_cost())?;
```
Comment (ahmadkaouk) [inline operator/precompiles/datahaven-native-transfer/src/lib.rs]: why we are recording a db_write_gas_cost here ?

File: operator/precompiles/datahaven-native-transfer/src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,228 @@
...
+#[cfg(test)]
+mod tests;
+
+type BalanceOf<Runtime> =
+    <<Runtime as pallet_datahaven_native_transfer::Config>::Currency as Inspect<
+        <Runtime as frame_system::Config>::AccountId,
+    >>::Balance;
+
+/// Storage read size for Paused storage value (bool = 1 byte + overhead)
+pub const PAUSED_STORAGE_SIZE: usize = 32;
```
Comment (ahmadkaouk) [inline operator/precompiles/datahaven-native-transfer/src/lib.rs]: It's actually 33.

```
  - Key = pallet prefix (Twox128, 16 B) + storage prefix (Twox128, 16 B) → 32 bytes.
  - Value = SCALE‐encoded bool → 1 byte.
  - Raw key+value payload = 32 + 1 = 33 bytes.
```

File: operator/precompiles/datahaven-native-transfer/src/lib.rs [hunk 11]
```diff
@@ -0,0 +1,228 @@
...
+            .map_err(|_| RevertReason::custom("Fee overflow").in_field("fee"))?;
+
+        // Validate amounts are non-zero
+        if amount_balance.into() == U256::zero() {
+            return Err(revert("Amount must be greater than zero"));
+        }
+
+        if fee_balance.into() == U256::zero() {
+            return Err(revert("Fee must be greater than zero"));
+        }
```
Comment (ahmadkaouk) [inline operator/precompiles/datahaven-native-transfer/src/lib.rs:112]: These checks are redundant. The pallet already enforces these (and more). It’s not harmful and can save a bit of gas on early failure, but it does create two sources of truth: if pallet validation changes, the precompile would still reject.

Comment (gonzamontiel) [reply operator/precompiles/datahaven-native-transfer/src/lib.rs:112]: Yeah I thought of that, but shouldn't add tests be enough for maintaning consistency ?

File: operator/precompiles/datahaven-native-transfer/src/lib.rs (hunk 10 above)
Comment (gonzamontiel) [reply operator/precompiles/datahaven-native-transfer/src/lib.rs]: I think this was misunderstood. The value is used for computing the db read cost, which for a boolean should be 1 byte. This value was being over conservative by trying to read a whole db word of 32 bytes. 
Looking at other precompile examples in Moonbeam (like https://github.com/moonbeam-foundation/moonbeam/blob/ff24ffea8649ed91fff837f3f674b28cc6aefb5c/precompiles/balances-erc20/src/lib.rs#L196) they seem to compute only for the SCALE-encoded value, in this case 1 byte. So I'll change it in the code to be 1 byte.

File: operator/precompiles/datahaven-native-transfer/src/lib.rs (hunk 5 above)
Comment (gonzamontiel) [reply operator/precompiles/datahaven-native-transfer/src/lib.rs]: Removed.

File: operator/precompiles/datahaven-native-transfer/src/lib.rs (hunk 7 above)
Comment (gonzamontiel) [reply operator/precompiles/datahaven-native-transfer/src/lib.rs]: my bad, left over from testing

File: operator/precompiles/datahaven-native-transfer/src/lib.rs (hunk 8 above)
Comment (gonzamontiel) [reply operator/precompiles/datahaven-native-transfer/src/lib.rs]: also a leftover, is not necessary, since try_dispatch will account already for storage reads and we don't perform additional storage reads here

File: operator/precompiles/datahaven-native-transfer/src/lib.rs (hunk 9 above)
Comment (gonzamontiel) [reply operator/precompiles/datahaven-native-transfer/src/lib.rs]: removed.

File: operator/precompiles/datahaven-native-transfer/src/lib.rs (hunk 10 above)
Comment (ahmadkaouk) [reply operator/precompiles/datahaven-native-transfer/src/lib.rs]: We actually account for the storage prefix and pallet prefix when computing proof sizes for storage reads. To verify a Merkle proof for a storage read, we need:
- The full storage key (to locate the node in the state trie): `twox128(pallet_name) ++ twox128(storage_name) ++ hasher(key)`
- The value (the actual data)

Without the pallet and storage prefixes, we couldn't reconstruct the full key and verify the proof is valid. Here is a valid example:
https://github.com/datahaven-xyz/datahaven/blob/0618e84268c82e4a794ccfbc2308d3af786aa9a1/operator/precompiles/collective/src/lib.rs#L42-L44

Speaking of this, since we’ve now set the `GasLimitPovSizeRatio` to 0, this value doesn’t hold much importance anymore.

File: operator/precompiles/datahaven-native-transfer/src/lib.rs (hunk 11 above)
Comment (gonzamontiel) [reply operator/precompiles/datahaven-native-transfer/src/lib.rs:112]: So, we already have tests that revert in this corner cases (zero address and amounts), the overflow check is necessary for the try_into(), so I think we should keep them as they don't validate some specific logic, but corner cases (transfer a zero amount is never a use case)

File: operator/precompiles/datahaven-native-transfer/DataHavenNativeTransfer.sol (hunk 6 above)
Comment (stiiifff) [reply operator/precompiles/datahaven-native-transfer/DataHavenNativeTransfer.sol]: @gonzamontiel has this comment been addressed ?

Comment (gonzamontiel) [reply operator/precompiles/datahaven-native-transfer/DataHavenNativeTransfer.sol]: Adressed now. I deleted the events we cannot record the log for, and added the record log for TockensLocked and TokensTransferredToEthereum
